homeassistant/components/template/template_entity.py

```diff
@@ -430,3 +445,220 @@ async def async_added_to_hass(self) -> None:
+                    continue
+
+                self._attr_extra_state_attributes.update(
+                    {key: last_sensor_state.attributes[key]}
+                )
+                _LOGGER.debug(
+                    "Restored attribute %s for entity %s to value %s",
+                    key,
+                    self.entity_id,
+                    value,
+                )
+
+        # Restore extra data
+        if (last_sensor_data := await self.async_get_last_template_data()) is None:
+            _LOGGER.debug(
+                "No extra data found to restore for entity %s", self.entity_id
+            )
+            return last_sensor_state, None
+
+        for attribute in self.additional_data:
+            try:
+                value = last_sensor_data[attribute]
+            except KeyError:
+                _LOGGER.debug(
+                    "No value stored for additional attribute %s for entity %s",
+                    attribute,
+                    self.entity_id,
+                )
+                continue
+
+            setattr(self, attribute, value)
+            _LOGGER.debug(
+                "Additional attribute %s restored to value %s for entity %s",
+                attribute,
+                value,
+                self.entity_id,
+            )
+
+        try:
+            self.async_write_ha_state()
+        except (TypeError, ValueError) as exc:
+            # Writing state resulted in an issue. Stop storing states for now.
+            self._save_state = False
+            _LOGGER.error(
+                "Restored state for entity %s results in exception: %s",
+                self.entity_id,
+                exc,
+            )
+
+        return (last_sensor_state, last_sensor_data)
+
+    @property
+    def extra_restore_state_data(self) -> TemplateExtraStoredData | None:
+        """Return sensor specific state data for persistent saving."""
+        return (
+            TemplateExtraStoredData(self, self.additional_data)
+            if self.restore
+            else TemplateExtraStoredData(self, [])
+        )
+
+    async def async_get_last_template_data(self) -> dict[str, Any] | None:
+        """Return the persistent saved data."""
+
+        if (restored_last_extra_data := await self.async_get_last_extra_data()) is None:
+            return None
+        return TemplateExtraStoredData(self, self.additional_data).from_dict(
+            restored_last_extra_data.as_dict()
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+
+        self._add_all_template_attributes()
+
+        await self.restore_entity()
+
+        if self.hass.state == CoreState.running:
+            await self._async_template_startup()
+            return
+
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, self._async_template_startup
+        )
+
+
+@dataclass
+class TemplateExtraStoredData(ExtraStoredData):
```

Comment:
Why do we need this?
See also https://github.com/home-assistant/core/pull/69350

Reply:
I developed this solution not just for template sensors but for all template platforms (sensor, binary sensor, select, number, weather, vacuum, ...). Some of these have additional items that should be restored that are not part of the attributes of the entity.
I originally created a PR that provided this for all template platforms but due to the size was asked to create a PR for 1 platform first. See PR #68713